### PR TITLE
fix: checkout base branch after commit files

### PR DESCRIPTION
**Pull request by k-jingyang**
## Changes

This line change gets the actual current branch from git instead of using the global variable `config.currentBranch` in `lb/util/git/index.ts`.

## Context

The bug occurs when two PRs have overlapping changes. E.g.

Focus on `org.springframework.boot``

1st PR
```diff
-	id 'org.springframework.boot'           version '2.2.5.RELEASE' apply false
+	id 'org.springframework.boot'           version '3.0.3' apply false 
```

2nd PR
```diff
-	id 'org.springframework.boot'          version '2.2.5.RELEASE'  apply false
-	id "com.github.node-gradle.node" 	version "3.2.1"        	         apply false
+	id 'org.springframework.boot'          version '2.7.9'                   apply false
+	id "com.github.node-gradle.node" 	version "3.5.1"        	         apply false
```

After renovate processes the 1st PR and is processing the 2nd PR, the current branch of the git repo is the branch of the previous PR whereas `config.currentBranch` reflects the base branch. When processing the 2nd PR, `isBranchConflicted` purpose is to check if the 2nd PR branch has a merge conflict with the base branch. But due to `isBranchConflicted`'s current implementation,

```typescript
const origBranch = config.currentBranch
try {
  await git.reset(ResetMode.HARD);
  //TODO: see #18600
  if (origBranch !== baseBranch) {
    await git.checkout(baseBranch);
  }
  await git.merge(['--no-commit', '--no-ff', `origin/${branch}`]);
```

In this case, origBranch == config.currentBranch == baseBranch while our actual current branch in git is on the previous PR. When we do the `git.merge` check, instead of comparing the 2nd PR against the base branch, we're actually comparing the 1st PR against the 2nd PR. Hence this causes the a merge conflict due to overlapping lines).

Related issue: #17804

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [x] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

Tested against this [repo](https://github.com/k-jingyang/renovate-test-17804)

### CODE REVIEW

**Comment by RahulGautamSingh** [comment]
Did you figure out how the baseBranch gets assigned with `config.currentBranch` ? I have been unable to find that.

**Comment by RahulGautamSingh** [comment]
For passing the semanticPR test, please change the Pr-title to something like:
 fix: use git to get current branch in `isBranchConflicted`

**Comment by k-jingyang** [comment]
> Did you figure out how/where the baseBranch gets assigned with `config.currentBranch` ? I have been unable to find that.

It was set once in [`syncGit()`](https://github.com/renovatebot/renovate/blob/main/lib/util/git/index.ts#L468). While processing the 1st PR, after the checkout in [prepareCommit](https://github.com/renovatebot/renovate/blob/main/lib/util/git/index.ts#L922), `config.currentBranch` and `config.currentBranchSha` was not set. 

We could also add the following to `prepareCommit` but I'm not sure about the implications on existing assumptions.

```diff
    await gitRetry(() =>
      git.checkout(['-B', branchName, 'origin/' + config.currentBranch])
    );
+  config.currentBranch = branchName;
    ...
+  config.currentBranchSha = commitSha;
    return result;
```

Regardless, using `git` to identify the current branch for conflict check should still be the way to go.

Thanks for taking a look at this! :smile: Could I also ask for this change to be ported to https://github.com/mend/renovate-on-prem once the PR is merged?

**Comment by RahulGautamSingh** [comment]
I am thinking that making changes to `prepareCommit` make more sense.

BTW
I introduced this bug in [this ](https://github.com/renovatebot/renovate/pull/16246/files)PR 🤦‍♂️ when I moved the `checkoutBranch(baseBranch)` after validation but the git-checks got affected by it.

**Comment by k-jingyang** [comment]
Woah, good catch! Can I insist on using
```typescript
  const origBranch = (await git.branch()).current;
```

as readers need not comprehend or rely on a non-local variable.

File: lib/workers/repository/update/branch/index.ts [hunk 1]
```diff
@@ -498,6 +498,9 @@ export async function processBranch(
       }
 
       commitSha = await commitFilesToBranch(config);
+      // Checkout to base branch to ensure that the next branch processing always starts with git being on the baseBranch
+      // baseBranch is not checked out at the start of processBranch() due to pull/16246
+      await checkoutBranch(config.baseBranch);
```
Comment (rarkins) [inline lib/workers/repository/update/branch/index.ts:503]: This line alone is not enough to fix the problem?

Comment (k-jingyang) [reply lib/workers/repository/update/branch/index.ts:503]: Yes, it'll be enough.

Comment (RahulGautamSingh) [reply lib/workers/repository/update/branch/index.ts:503]: Then we can revert the others after running on a test repo.

Comment (k-jingyang) [reply lib/workers/repository/update/branch/index.ts:503]: Sure, reverted them.

**Review by RahulGautamSingh — COMMENTED** [review]
LGTM 👍

**Review by viceice — COMMENTED** [review]
needs test changes

**Comment by k-jingyang** [comment]
> needs test changes

What do you mean?

**Comment by RahulGautamSingh** [comment]
> > needs test changes
> 
> What do you mean?

After making change to a file, we check if it effects the existing tests written for that file. Or we need a new test that could help us be certain that the changes are okay. In this case the file will be `lib/workers/repository/update/branch/index.ts`

To test locallly run `yarn jest lib/workers/repository/update/branch/index.ts` this can be shortened since its regex matching.

**Comment by rarkins** [comment]
Here's where we check out branches today:
image

Here are the ones relating to `baseBranch`:

Before extracting: https://github.com/renovatebot/renovate/blob/56c161a8d4bc8ebf661a641fdcd12d803492749f/lib/workers/repository/process/extract-update.ts#L140-L142

Before checking for updated package files:
https://github.com/renovatebot/renovate/blob/56c161a8d4bc8ebf661a641fdcd12d803492749f/lib/workers/repository/update/branch/index.ts#L384-L386

Before branch automerge: https://github.com/renovatebot/renovate/blob/56c161a8d4bc8ebf661a641fdcd12d803492749f/lib/workers/repository/update/branch/automerge.ts#L37-L46

So this will add a fourth one - after committing any files?

**Comment by RahulGautamSingh** [comment]
Yes

**Comment by k-jingyang** [comment]
I added a test for it, and have verified that without the added `await checkoutBranch(config.baseBranch);`, the test will fail.

I'm not familiar with Jest, so do let me know if there are better APIs to use for the test.

**Review by viceice — APPROVED** [review]
🤷‍♂️

**Pull request merged**